### ISSUE TITLE
No accessToken generated in android

### ISSUE THREAD

**Issue by Barathwaja** [EddyVerbruggen/cordova-plugin-googleplus#337]
Hi there i was able to get idToken and serverAuthToken but not accessToken.. while searching in GooglePlus.java there is nothing written to call accessToken right?? but for iOS its there...

**Comment by hughred22** [comment]
I am running into the same issue. On iOs I can call YouTube immediately in the signin function as I have accessToken but not in Android!! Please help!

**Comment by eye4brain** [comment]
The original cordova-plugin-googleplus can get what is called idToken, but it seems that you can not get accessToken required by Google API.

Since cordova-plugin-googleplus forked by **[Mr. yhwh](https://github.com/yhwh/cordova-plugin-googleplus/blob/master/src/android/GooglePlus.java)** contains the getAuthUtil method, it can be acquired.

There seems to be a way to get **[accesstoken from idToken](https://developers.google.com/identity/protocols/OAuth2ServiceAccount)**, but I could not get it successfully by putting urn: ietf: params: oauth: grant - type: jwt - bearer in the grant - type.

By inviting you to include the accessToken acquisition code in the original cordova-plugin-googleplus (because it seems that oauth 2 authentication **[can not be done](https://developers.googleblog.com/2016/08/modernizing-oauth-interactions-in-native-apps.html)** in webview nearby)

I would be happy if you could see the **[page](http://wp.me/p5KJV2-164)** you summarized if you enjoy it. It is Japanese.

**Comment by hughred22** [comment]
OMG! Can someone translate the Japanese version. This can be the exact info I need if I can read that~~~~

**Comment by sammuggPS** [comment]
The idToken cannot be used to get an access or refresh token. All it is is a JWT containing info that identifies (authenticates) a user. It cannot be used for authorizing Google API Access.

The serverAuthCode (or authorization code) is what is used to get an access and/or refresh token.

https://developers.google.com/identity/protocols/OAuth2InstalledApp

**Comment by Barathwaja** [comment]
I was actually able to do like Client -> My Server -> Google Server communication and able to get key, I used this combo for my scenario.. But if **_iOS_** is able to get **access token** why cant **_android_** too ??? @sammuggPS ... why @EddyVerbruggen wrote only for iOS ??

**Comment by sammuggPS** [comment]
It's not because of anything that was written for this plugin.

The Google SDK is deeply integrated into Android as a platform. That's why the login experience is different on Android devices than iOS. If you're logging a user in on Android, Google wants you to use the native SDKs for all API integrations (i.e. native (java) code).

However, iOS does not benefit from having Google SDKs deeply embedded into their software. Therefore, they always send back an accessToken when a user gets logged in because it is required for continued Google API use in iOS applications.

**Comment by hughred22** [comment]
Since we are using Ionic 2 not native java code. How can we get the accessToke just like iOs so we can continue the integration? The react Native google oauth plugin has this build in for Android so both iOs and Android can return the same info so we can call google immediate after login to get more user info - like in my case - their YouTube channel info to populate my app. But without accessToken in Android, my app won't work on Android (only in web browser and iOS)

**Comment by Barathwaja** [comment]
I know they focus on **Android** platform , but still consider my **_first scenario_** what if the access token ( **_iOS  platform_** ) got exposed and that leads to vulnerability and finally exposed to data's due to single token?? google server gets impacted?? **_Another scenario_** is in Android I need to call Google Server to get access token and inturn have to use that to call API's leading to my time wastage instead in iOS already token is generated then everyone can deploy their application in iOS and make work easier rite?? 

But as far as I've googled in _GOOGLE API_ there is nothing mentioned that iOS will get token and Android will not and all ... They considered both platforms equal and gave their suggestion on how to get access token ... @sammuggPS

**Comment by sammuggPS** [comment]
Keep in mind that Google Sign-In is designed for use with Native applications. Not Hybrid apps.

As a result, we as developers need to make design decisions in how we integrate Native authentication into our hybrid apps. The most common design decision is to exchange the serverAuthCode for an access token. The authCode is available on both Android and iOS.

Yes there is some extra work to get things configured and work the way we want them to, but that's the cost of using these technologies together.

**Comment by hughred22** [comment]
@sammuggPS Agree. But could you give us an example code on how to do that in JavaScript to exchange serverAuthCode to accessToken? PLEASE if I ask super nicely :) I think want a pure front-end JavaScript solution without server involved. As this is going to be built into an native app. Security is not my concern if user can see my client ID or what not. I just want to get the accessToken without server call (only calling Google server). Thanks!

**Comment by sammuggPS** [comment]
I don't have the bandwidth to write up an example at this time, but everything you need should be in Google's documentation. Try here for starters: https://developers.google.com/identity/protocols/OAuth2WebServer#exchange-authorization-code

**Comment by sammuggPS** [comment]
You should be able to do it with a simple REST or curl call

**Comment by hughred22** [comment]
Hi, I am confused with which is Client Secret. The google docs said "The client secret obtained from the API Console." But I can not find it in the API Console. The Android key does not have any secret... Is the Client Secret refer to the Web Application Client Secret?

**Comment by sammuggPS** [comment]
I believe so

**Comment by johannesmauerer** [comment]
There is a fork that includes accessToken for Android: https://github.com/yhwh/cordova-plugin-googleplus/blob/master/src/android/GooglePlus.java#L328

**Comment by hughred22** [comment]
@johannesmauerer I install the fork but I still does not get the accessToken in Android. No idea what is going wrong...

**Comment by johannesmauerer** [comment]
@hughred22 how are you invoking the flow? Do you have code examples?

**Comment by Barathwaja** [comment]
@hughred22 Check the below picture for testing via POSTMAN.

The process is you will get _client_id_ and _client_secret_ from the **JSON** file , while the **code** there is the _server access token_ which you will get from the ionic android after loggin in and validating with _google server_,make sure the grant_type is **authorization_code**

The response will be _accesstoken_ with which you can access _Google API's by Enabling them_, along with that you will get _idToken_ (JWT), _refresh token_ - which will be our **next server code**, and _expiry_ ..


![google_oauth](https://cloud.githubusercontent.com/assets/5066436/24340135/97d710b2-12cd-11e7-8eb5-99b8dcf0972f.png)

**Comment by CoverID** [comment]
thank you @johannesmauerer , https://github.com/yhwh/cordova-plugin-googleplus/blob/master/src/android/GooglePlus.java#L328 solved my issue

**Comment by theunreal** [comment]
@CoverID How did u add it to ur project?

**Comment by Sampath-Lokuge** [comment]
@CoverID Here is the solution: https://github.com/EddyVerbruggen/cordova-plugin-googleplus/issues/398#issuecomment-316670140

**Comment by Ebraheemb** [comment]
> It's not because of anything that was written for this plugin.
> 
> The Google SDK is deeply integrated into Android as a platform. That's why the login experience is different on Android devices than iOS. If you're logging a user in on Android, Google wants you to use the native SDKs for all API integrations (i.e. native (java) code).
> 
> However, iOS does not benefit from having Google SDKs deeply embedded into their software. Therefore, they always send back an accessToken when a user gets logged in because it is required for continued Google API use in iOS applications.

This answer was very helpful for me thank you so much

**Comment by johannesjo** [comment]
Did anybody find a way to get the access token from android. I spend days on this and I feel like I am loosing my mind.

**Comment by emiliam** [comment]
Please merge https://github.com/CodetrixStudio/CapacitorGoogleAuth/pull/56/files to fix this for us all.
